rdkit/Chem/PandasTools.py

```diff
@@ -169,48 +170,96 @@ def _getPandasVersion():
 
 
 def patchPandasHTMLrepr(self, **kwargs):
-    '''
-    Patched default escaping of HTML control characters to allow molecule image rendering dataframes
-    '''
-    # set escape to False if not set
-    kwargs['escape'] = kwargs.get('escape') or False
+    """A patched version of the DataFrame.to_html method that allows rendering
+    molecule images in data frames.
+    """
+
+    # Two things have to be done:
+    # 1. Disable escaping of HTML in order to render img / svg tags
+    # 2. Avoid truncation of data frame values that contain HTML content
 
-    # do not allow pandas to truncate text since PNG byte strings are lengthy
-    with pd.option_context('display.max_colwidth', -1):
+    # 1. Temporarily set escape=False in HTMLFormatter._write_cell
+    import pandas.io.formats.html  # necessary for loading HTMLFormatter
+    defHTMLFormatter_write_cell = pd.io.formats.html.HTMLFormatter._write_cell
+
+    def patched_HTMLFormatter_write_cell(self, s, *args, **kwargs):
+        def_escape = self.escape
+        try:
+            if is_molecule_image(s):
+                self.escape = False
+            return defHTMLFormatter_write_cell(self, s, *args, **kwargs)
+        finally:
+            self.escape = def_escape
+
+    # 2. Pandas uses TextAdjustment objects to measure the length of texts
+    #    (e.g. for east asian languages). We take advantage of this mechanism
+    #    and replace the original text adjustment object with a custom one.
+    #    This "RenderMoleculeAdjustment" object assigns a length of 0 to a
+    #    given text if it is valid HTML. And a value having length 0 will not
+    #    be truncated.
+
+    # store original _get_adjustment method
+    defPandasGetAdjustment = pd.io.formats.format._get_adjustment
+
+    def patched_get_adjustment():
+        inner_adjustment = defPandasGetAdjustment()
+        return RenderMoleculeAdjustment(inner_adjustment)
+
+    try:
+        # patch _get_adjustment method and call original to_html function
+        pd.io.formats.format._get_adjustment = patched_get_adjustment
+        pd.io.formats.html.HTMLFormatter._write_cell = patched_HTMLFormatter_write_cell
         return defPandasRendering(self, **kwargs)
```

Comment:
I worry that pandas might change under the hood, what happens if an exception is thrown?

Perhaps we should restore the patches, and call defPandasRendering again in this case.

I might also suggest:

```
if getattr(pd.io.formats, "_get_adjustmant"):
```

which might future protect this code.

Reply:
Hey @bp-kelley,

thanks for your feedback! I implemented a fallback for the case that these specific private methods in pandas are moved or deleted. And if an exception is thrown, the old patch is used (avoids truncation and disables escaping on the whole data frame). This old patch uses only public pandas methods (e.g. should not raise errors) and it is likely that we still see molecule images in the rendered data frames.